### PR TITLE
Automatic allocation of compound constitutive relations 

### PR DESCRIPTION
**Pull request by CusiniM**
close issue #1528 

- [x] implement allocation
- [x] modify multiphysics xmls
- [x] modify fluidFlow xmls.
- [x] rebaseline integratedTests (https://github.com/GEOSX/integratedTests/pull/158)

### CODE REVIEW

**Comment by CusiniM** [comment]
I am only not sure about what will happen if someone specifies both the compound and the standalone model on the same Region. Will it work? Should we throw an error?

File: src/coreComponents/constitutive/ConstitutiveManager.cpp [hunk 1]
```diff
@@ -67,6 +67,19 @@ ConstitutiveManager::hangConstitutiveRelation( string const & constitutiveRelati
...
                                       numConstitutivePointsPerParentIndex );
 
+
+  std::vector<string> const subRelationNames = material->getSubRelationNames();
+
+  for( string const & subRelationName : subRelationNames )
+  {
+    ConstitutiveBase const & subRelation = getConstitutiveRelation( subRelationName );
+
+    std::unique_ptr< ConstitutiveBase > constitutiveModel = subRelation.deliverClone( subRelationName, parent );
```
Comment (klevzoff) [inline src/coreComponents/constitutive/ConstitutiveManager.cpp:77]: `constitutiveModel` goes out of scope and destructs at the end of the loop body. It needs to be moved a wrapper on the new `constitutiveGroup`, similar to what happens to `material` some lines below.

Comment (CusiniM) [reply src/coreComponents/constitutive/ConstitutiveManager.cpp:83]: should work fine now. Should I add a check that if the wrapper already exists we don't register it?

File: src/coreComponents/constitutive/ConstitutiveManager.cpp [hunk 2]
```diff
@@ -60,25 +60,42 @@ ConstitutiveManager::hangConstitutiveRelation( string const & constitutiveRelati
...
+      group = constitutiveGroup->registerGroup< ConstitutiveBase >( subRelationName, std::move( constitutiveModel ) );
+      group.setSizedFromParent( 1 );
+      group.resize( constitutiveGroup->size() );
+  }
 
   ConstitutiveBase &
   rval = constitutiveGroup->registerGroup< ConstitutiveBase >( constitutiveRelationInstanceName, std::move( material ) );
   rval.setSizedFromParent( 1 );
   rval.resize( constitutiveGroup->size() );
+  // Why do we need to return this here? Couldn't this be just void?
```
Comment (CusiniM) [inline src/coreComponents/constitutive/ConstitutiveManager.cpp:98]: I was wondering why we need this function to return that reference. It does not seem to be used.

**Comment by sytuannguyen** [comment]
Please modify also the multiphysics/benchmark XMLs

File: src/coreComponents/constitutive/ConstitutiveManager.cpp [hunk 3]
```diff
@@ -55,33 +55,59 @@ void ConstitutiveManager::expandObjectCatalogs()
...
-  ConstitutiveBase const & constitutiveRelation = getConstitutiveRelation( constitutiveRelationInstanceName );
-
-  std::unique_ptr< ConstitutiveBase > material = constitutiveRelation.deliverClone( constitutiveRelationInstanceName, parent );
-
-  material->allocateConstitutiveData( *parent,
-                                      numConstitutivePointsPerParentIndex );
-
   dataRepository::Group * constitutiveGroup = parent->getGroupPointer( groupKeyStruct::constitutiveModelsString() );
-  if( constitutiveGroup == nullptr )
+  if( constitutiveGroup == nullptr )   // Why do we need to do this check? shouldn't it always be there?
```
Comment (klevzoff) [inline src/coreComponents/constitutive/ConstitutiveManager.cpp:64]: The way we currently use this by passing a descendant of `ElementSubRegionBase` - yes, it always has a constitutive subgroup in it. Bit the function signature takes an arbitrary `Group`, which may or may not have it. If we dropped the check, someone could unwittingly try to put constitutive models on something that is not a subregion and get segfaults. We could change the function signature to take `ElementSubRegionBase`, but that would require taking a dependency on `mesh` which I believe is not desirable and possibly circular.

File: src/coreComponents/constitutive/ConstitutiveManager.cpp [hunk 4]
```diff
@@ -55,33 +55,59 @@ void ConstitutiveManager::expandObjectCatalogs()
...
+    ConstitutiveBase &
+    rval = constitutiveGroup->registerGroup< ConstitutiveBase >( constitutiveRelationInstanceName, std::move( material ) );
+    rval.setSizedFromParent( 1 );
+    rval.resize( constitutiveGroup->size() );
+
+    // 2. Allocate subrelations (for compound models)
+    std::vector< string > const subRelationNames = constitutiveRelation.getSubRelationNames();
+    for( string const & subRelationName : subRelationNames )
+    {
+      std::cout << "nothing to do" << std::endl;
```
Comment (klevzoff) [inline src/coreComponents/constitutive/ConstitutiveManager.cpp:90]:
```suggestion
```

File: src/coreComponents/constitutive/ConstitutiveManager.cpp [hunk 5]
```diff
@@ -55,33 +55,59 @@ void ConstitutiveManager::expandObjectCatalogs()
...
+      // we only register the group if the instance has not been registered yet.
+      if( !constitutiveGroup->hasGroup( subRelationName ) )
+      {
+        ConstitutiveBase &
+        group = constitutiveGroup->registerGroup< ConstitutiveBase >( subRelationName, std::move( constitutiveModel ) );
+        group.setSizedFromParent( 1 );
+        group.resize( constitutiveGroup->size() );
+      }
+    }
+    std::cout << "done" << std::endl;
```
Comment (klevzoff) [inline src/coreComponents/constitutive/ConstitutiveManager.cpp:107]:
```suggestion
```

File: src/coreComponents/constitutive/ConstitutiveManager.cpp [hunk 6]
```diff
@@ -55,33 +55,59 @@ void ConstitutiveManager::expandObjectCatalogs()
...
+      std::cout << "nothing to do" << std::endl;
+      ConstitutiveBase const & subRelation = getConstitutiveRelation( subRelationName );
+
+      std::unique_ptr< ConstitutiveBase > constitutiveModel = subRelation.deliverClone( subRelationName, parent );
+
+      constitutiveModel->allocateConstitutiveData( *parent,
+                                                   numConstitutivePointsPerParentIndex );
+
+      // we only register the group if the instance has not been registered yet.
+      if( !constitutiveGroup->hasGroup( subRelationName ) )
```
Comment (klevzoff) [inline src/coreComponents/constitutive/ConstitutiveManager.cpp:99]: I am not 100% certain, but if the sub-relation has already been registered (presumably by another compound model, or explicitly by the user by listing it in region's material list), should we also avoid calling `deliverClone` and `allocateConstitutiveData` above, i.e. move entire loop body under this branch?

**Review by klevzoff — APPROVED** [review]
Nice! The XMLs are looking leaner now.

File: src/coreComponents/constitutive/ConstitutiveManager.cpp (hunk 6 above)
Comment (CusiniM) [reply src/coreComponents/constitutive/ConstitutiveManager.cpp:98]: yes, I think you are right. I don't see why we would not want to do it.

File: src/coreComponents/constitutive/ConstitutiveManager.cpp [hunk 7]
```diff
@@ -55,33 +55,58 @@ void ConstitutiveManager::expandObjectCatalogs()
...
   dataRepository::Group * constitutiveGroup = parent->getGroupPointer( groupKeyStruct::constitutiveModelsString() );
   if( constitutiveGroup == nullptr )
   {
     constitutiveGroup = &parent->registerGroup( groupKeyStruct::constitutiveModelsString() ).setSizedFromParent( 1 );
     constitutiveGroup->resize( parent->size() );
   }
 
+  // 1. Allocate constitutive relation
+  // we only register the constitutive relation if it has not been registered yet.
+  if( !constitutiveGroup->hasGroup( constitutiveRelationInstanceName ) )
```
Comment (rrsettgast) [inline src/coreComponents/constitutive/ConstitutiveManager.cpp:72]: hard error?

File: src/coreComponents/constitutive/ConstitutiveManager.cpp [hunk 8]
```diff
@@ -55,33 +55,58 @@ void ConstitutiveManager::expandObjectCatalogs()
...
+    materialGroup = constitutiveGroup->registerGroup< ConstitutiveBase >( constitutiveRelationInstanceName, std::move( material ) );
+    materialGroup.setSizedFromParent( 1 );
+    materialGroup.resize( constitutiveGroup->size() );
 
+    // 2. Allocate subrelations (for compound models)
+    std::vector< string > const subRelationNames = constitutiveRelation.getSubRelationNames();
+    for( string const & subRelationName : subRelationNames )
+    {
+      // we only want to register the subRelation if it has not been registered yet.
+      if( !constitutiveGroup->hasGroup( subRelationName ) )
```
Comment (rrsettgast) [inline src/coreComponents/constitutive/ConstitutiveManager.cpp:91]: hard error?

**Review by rrsettgast — APPROVED** [review]

File: src/coreComponents/constitutive/ConstitutiveManager.cpp (hunk 7 above)
Comment (CusiniM) [reply src/coreComponents/constitutive/ConstitutiveManager.cpp]: added.

File: src/coreComponents/constitutive/ConstitutiveManager.cpp (hunk 8 above)
Comment (CusiniM) [reply src/coreComponents/constitutive/ConstitutiveManager.cpp]: added.

File: src/coreComponents/constitutive/ConstitutiveManager.cpp [hunk 9]
```diff
@@ -55,33 +55,64 @@ void ConstitutiveManager::expandObjectCatalogs()
 }
 
 
-ConstitutiveBase &
+void
 ConstitutiveManager::hangConstitutiveRelation( string const & constitutiveRelationInstanceName,
                                                dataRepository::Group * const parent,
```
Comment (sytuannguyen) [inline src/coreComponents/constitutive/ConstitutiveManager.cpp:60]:
```suggestion
                                               Group * const parent,
```

File: src/coreComponents/constitutive/ConstitutiveManager.cpp [hunk 10]
```diff
@@ -55,33 +55,64 @@ void ConstitutiveManager::expandObjectCatalogs()
...
                                                localIndex const numConstitutivePointsPerParentIndex ) const
 {
+  dataRepository::Group * constitutiveGroup = parent->getGroupPointer( groupKeyStruct::constitutiveModelsString() );
```
Comment (sytuannguyen) [inline src/coreComponents/constitutive/ConstitutiveManager.cpp:63]:
```suggestion
  Group * constitutiveGroup = parent->getGroupPointer( groupKeyStruct::constitutiveModelsString() );
```

File: src/coreComponents/constitutive/ConstitutiveManager.hpp [hunk 11]
```diff
@@ -52,7 +52,7 @@ class ConstitutiveManager : public dataRepository::Group
   /// This function is used to expand any catalogs in the data structure
   virtual void expandObjectCatalogs() override;
 
-  ConstitutiveBase &
+  void
   hangConstitutiveRelation( string const & constitutiveRelationInstanceName,
                             dataRepository::Group * const parent,
```
Comment (sytuannguyen) [inline src/coreComponents/constitutive/ConstitutiveManager.hpp:57]:
```suggestion
                            Group * const parent,
```

**Review by sytuannguyen — APPROVED** [review]

**Pull request merged**